Replace http-proxy with http-proxy-3

StorytellerCZ opened the pull request:
http-proxy is not maintained, has security & performance concerns and uses deprecated API which throws when you start Meteor: #13491
I have found that [http-proxy-3](https://www.npmjs.com/package/http-proxy-3) is the most recent replacement package and went ahead and replaced it in the core. My app runs without the deprecation warning on the Meteor 3.4 branch, so I'm putting it here to run with the other tests (though I think new bundle might be necessary).
I think http-proxy-3 is even better option than forking it ourselves as was suggested in related discussions.

StorytellerCZ committed:
Replace http-proxy with http-proxy-3

StorytellerCZ commented:
@italojs @Grubba27 @nachocodoner
This will have to be part of a new bundle. For Meteor 3.4 should I create Node 22 or push this into the Node 24 bundle?

nachocodoner commented:
We plan to add Node 24 to 3.4. We’re not sure yet if it will cause issues, but I expect the migration to go smoothly. We’ll verify soon, so you can either prepare this PR for Node 24 changes now or wait until we confirm. In any case, we should keep the this http-proxy-3 upgrade and the Node upgrade in separate PRs to focus on each change verifications.

nachocodoner committed:
Merge branch 'devel' into feature/http-proxy-3

# Conflicts:
#	package.json

nachocodoner committed:
Merge branch 'release-3.4' into feature/http-proxy-3

# Conflicts:
#	package.json

nachocodoner committed:
Update BUNDLE_VERSION to 22.18.0.100 to test http-proxy-3

StorytellerCZ commented:
@nachocodoner this can be merged into the branch and it should work with any of the new dev bundles. Just need to make sure to build the bundle. Should I change the merge target to the Node 24 dev bundle or keep it as is?

nachocodoner commented:
> Should I change the merge target to the Node 24 dev bundle or keep it as is?

Refer to this comment: https://github.com/meteor/meteor/pull/13948#issuecomment-3415639492

I’d prefer to get a PR with the latest Node 22.x series for Meteor 3.4 instead.

I’m generating a new dev_bundle to run the checks on this PR and make sure they pass before merging. Once we get a green signal, we can move forward this change with a 3.4 beta release.

nachocodoner committed:
re-run checks

nachocodoner reviewed (changes requested):
It seems this PR breaks one test:

<img width="2576" height="189" alt="image" src="https://github.com/user-attachments/assets/8aa768a3-2fa9-44a4-bb0d-aa62e9c97de1" />

I will wait for it to be fixed to be included.

StorytellerCZ committed:
Merge branch 'release-3.4' into feature/http-proxy-3

StorytellerCZ committed:
Bump to http-proxy-3 v1.22.0

StorytellerCZ commented:
OK, so the error happens before the upgrade as well, but it fails after the test get a chance to run successfully.
```
node:fs:2168
  binding.chown(path, uid, gid);
          ^

Error: EPERM: operation not permitted, chown '/tmp/socket_file_tests'
    at chownSync (node:fs:2168:11)
    at exports.main (packages/webapp/webapp_server.js:1449:9)
    at Object.<anonymous> (packages/local-test:webapp/socket_file_tests.js:140:22) {
  errno: -1,
  code: 'EPERM',
  syscall: 'chown',
  path: '/tmp/socket_file_tests'
}
```

So not exactly related to this (this PR just makes it appear more prominently), but needs to be fixed none the less.

StorytellerCZ committed:
Fix usage of Meteor.absoluteUrl

StorytellerCZ committed:
Further fix query test

StorytellerCZ commented:
So I got it passing locally. Let's see what the CI says.

nachocodoner commented on packages/webapp/socket_file_tests.js:
```diff
@@ -140,7 +140,9 @@ testAsyncMulti(
-      test.equal((await getChownInfo(testSocketFile))?.gid, getGroupInfo(process.env.UNIX_SOCKET_GROUP)?.gid);
+      const currentGid = userInfo({ encoding: "utf8" })?.gid;
+      const expectedGid = process.getuid() === 0 ? getGroupInfo(process.env.UNIX_SOCKET_GROUP)?.gid : currentGid;
+      test.equal((await getChownInfo(testSocketFile))?.gid, expectedGid);
```
I wouldn’t recommend modifying the test implementation to make the test pass. Ideally, it should remain the same, and only the main implementation should be adapted to fix it.

nachocodoner commented:
Suddenly, more tests are failing. I only showed one failing test in my comment above. The recent commits likely made things worse.

The only test to fix was `webapp - parse url queries`. If you revert changes you will likely get back into the state where it was the only one to fix.

<img width="2576" height="189" alt="image" src="https://github.com/user-attachments/assets/1ee7830c-0f7d-464c-bbad-19d391ef837c" />

---

> So not exactly related to this (this PR just makes it appear more prominently),

I have not seen that test failing in other PRs even as flacky. Did you? Where did it fail?

StorytellerCZ committed:
Revert socket tests changes

StorytellerCZ commented:
Still investigating, but I think what is failing is actually outside of the actual changes I made, but they just appeared to make it visible.
For example when I reverted the changes in `socket_file_test.js` and `webapp_server.js` it is the socket usage test that now starts failing locally with the same error:
```
node:fs:2168
  binding.chown(path, uid, gid);
          ^

Error: EPERM: operation not permitted, chown '/tmp/socket_file_tests'
    at chownSync (node:fs:2168:11)
    at exports.main (packages/webapp/webapp_server.js:1449:9)
    at Object.<anonymous> (packages/local-test:webapp/socket_file_tests.js:140:22) {
  errno: -1,
  code: 'EPERM',
  syscall: 'chown',
  path: '/tmp/socket_file_tests'
}

Node.js v22.20.0
```

This is quiet outside of my wheelhouse so I will have to dig a bit more deeper here.

StorytellerCZ committed:
Restore privilege guard

StorytellerCZ commented:
So digging around and asking help from AI to help me explain what is going on and check my thoughts lead me to this understanding so far:
* Code intent: `webapp/main()` creates a UNIX socket at `process.env.UNIX_SOCKET_PATH`, then honors `process.env.UNIX_SOCKET_GROUP` by `chownSync`‑ing the socket so other processes can connect. The tests in `packages/webapp/socket_file_tests.js` assert that the gid actually changes to the target group (root on Linux when TRAVIS is falsy).
* OS constraint: On Linux, only a privileged process (uid `0`, or one with `CAP_CHOWN`) can change a file’s group to one it doesn’t belong to (root → gid `0`). When you run the package tests as your normal user (uid/gid `1000`), `chownSync` cannot succeed, so the gid remains `1000` and the assertion fails. The upgrade to `http-proxy-3` didn’t change this logic; it just exposed the permission requirement in environments where tests used to be executed with elevated privileges. Which is why when I made the code to fail more gracefully and speed things up a bit with better usage of `Meteor.absoluteUrl` the failure moved to the `use socket file for inter-process communication` test where it actually fails.
* `travis` and on MacOS `staff` will succeed this test, but not so on Linux where the user is not typically `root` (must invoke `sudo` for example), the test fails because of that.

I have added a check to skip the test on Linux and try/catch around `chownSync` to gracefully fail if we run into this.
@nachocodoner I have merged a bit back from the 3.4 branch so new bundle will be required for this and is probably the reason why the other tests failed. Apologies about that. 🙇 I have upped the patch version of the bundle in here so it should be straight forward to go with this and reveal the actual errors once again (though except the permission issue all the packages test have passed for me locally).

Now I have changed things according to the best of my understanding, but this is not my strong area so I can be totally wrong, so please pick this apart as this is an important area for Meteor.

StorytellerCZ committed:
Fallback for chownSync on Linux

nachocodoner commented:
I will move its milestone to 3.5 instead. We don't have capacity in Meteor 3.4 to finish this PR given everything else.

We're trying to focus on expanding the bundler, and every week we get new feedback to address. We'd like to reach the official release as soon as possible. To avoid more delays, I'll look deeper into this PR, changes, and tests later.

StorytellerCZ committed:
Merge branch 'release-3.4' into feature/http-proxy-3

StorytellerCZ committed:
Merge branch 'release-3.5' into feature/http-proxy-3

StorytellerCZ commented:
I'm coming back to this now that 3.4 is released.
Would be great to get the annoying warning removed from the console on start.
@nachocodoner @italojs any help or directions with the test are welcome.

dupontbertrand commented:
Cc @nachocodoner :eyes: it just needs a core team member to rebuild the dev_bundle and run CI imo ?

nachocodoner commented:
Yes. That is the very next step. Thanks for the reminder.

Before that, we needed to prepare the CI to work with Node 24, and that happened recently.

We will proceed with generating one for it as soon as we can. It is planned for Meteor 3.5 if everything goes well with CI and the testing period. We have many fronts open, including wrapping up 3.4.1, but we will get there as we can.

italojs committed:
Merge branch 'release-3.5' into feature/http-proxy-3

italojs commented:
the branch `release-3.5` already is working on node 24 LTS, maybe if we update the branch is enougth

nachocodoner commented:
> maybe if we update the branch is enougth

Update it for sure, but also generate a new dev_bundle, as it changes `scripts/dev-bundle-tool-package.js`. Otherwise the replacement doesn't get any effect on the CI test env.

nachocodoner committed:
bump `BUNDLE_VERSION` to 24.14.0.5

nachocodoner committed:
re-run checks

nachocodoner reviewed (dismissed):
Bumped and generated a new dev_bundle 24.14.0.5. Test seems to have passed.

italojs commented:
the test packages CI stucks and get canceled after 90min


it stucks while run `./packages/test-in-console/run.sh`

logs from CI
```
[...]
#######################################################################   99.5%
######################################################################## 100.0%
Installed dependency kit v24.14.0.5 in dev_bundle.
meteor: updating npm dependencies -- denque...
modules-runtime: updating npm dependencies -- install...
modules: updating npm dependencies -- @meteorjs/reify, meteor-babel-helpers...
react-fast-refresh: updating npm dependencies -- react-refresh, semver...
promise: updating npm dependencies -- promise...
ecmascript-runtime-client: updating npm dependencies -- core-js...
ecmascript-runtime-server: updating npm dependencies -- core-js...
ecmascript-runtime: updating npm dependencies -- @babel/runtime...
fetch: updating npm dependencies -- node-fetch, whatwg-fetch...
inter-process-messaging: updating npm dependencies -- uuid, arson...
babel-compiler: updating npm dependencies -- @meteorjs/babel, json5, semver,
@meteorjs/swc-core...
ecmascript: updating npm dependencies -- @babel/runtime...
socket-stream-client: updating npm dependencies -- faye-websocket,
permessage-deflate2, lodash.isequal, lodash.once...
ddp-client: updating npm dependencies -- @sinonjs/fake-timers, lodash.has,
lodash.identity...
npm-mongo: updating npm dependencies -- mongodb...
package-version-parser: updating npm dependencies -- semver...
logic-solver: updating npm dependencies -- lodash.has...
constraint-solver: updating npm dependencies -- lodash.has, lodash.memoize,
lodash.isequal, lodash.isempty, lodash.zip, lodash.groupby, lodash.isstring,
lodash.isobject...
boilerplate-generator: updating npm dependencies -- combined-stream2...
logging: updating npm dependencies -- chalk, @babel/runtime...
npm warn Unknown env config "nodedir". This will stop working in the next major version of npm. See `npm help npmrc` for supported config options.
npm warn deprecated puppeteer@23.6.0: < 24.15.0 is no longer supported
added 104 packages in 9s
10 packages are looking for funding
  run `npm fund` for details
(node:3479) Warning: `url.parse()` behavior is not standardized and prone to errors that have security implications. Use the WHATWG URL API instead. CVEs are not issued for `url.parse()` vulnerabilities.
(Use `node --trace-warnings ...` to show where the warning was created)
```

[edit 1]

logs from my machine, the error is -> ` Identifier 'groupToUse' has already been declared. (192:12)`
```
############################################################################################################################################################################# 100.0%
Installed dependency kit v24.14.0.5 in dev_bundle.

npm warn Unknown env config "nodedir". This will stop working in the next major version of npm. See `npm help npmrc` for supported config options.
npm warn deprecated puppeteer@23.6.0: < 24.15.0 is no longer supported

added 104 packages in 3s

10 packages are looking for funding
  run `npm fund` for details
(node:41839) Warning: `url.parse()` behavior is not standardized and prone to errors that have security implications. Use the WHATWG URL API instead. CVEs are not issued for `url.parse()` vulnerabilities.
(Use `node --trace-warnings ...` to show where the warning was created)
[[[[[ Tests ]]]]]

=> Started proxy.
=> Started MongoDB.
(node:41839) Warning: Accessing non-existent property 'lineno' of module exports inside circular dependency
(node:41839) Warning: Accessing non-existent property 'column' of module exports inside circular dependency
(node:41839) Warning: Accessing non-existent property 'filename' of module exports inside circular dependency
(node:41839) Warning: Accessing non-existent property 'lineno' of module exports inside circular dependency
(node:41839) Warning: Accessing non-existent property 'column' of module exports inside circular dependency
(node:41839) Warning: Accessing non-existent property 'filename' of module exports inside circular dependency
=> Errors prevented startup:
   
   While building for os.osx.arm64:
   packages/local-test:webapp/socket_file_tests.js:192:12:
   /Users/italojose/dev/meteor-dev/meteor/packages/local-test:webapp/socket_file_tests.js:
   Identifier 'groupToUse' has already been declared. (192:12)
   
   190 |
   191 |       // use UNIX_SOCKET_PATH and UNIX_SOCKET_GROUP
   > 192 |       const groupToUse = getWritableGroupName();
   |             ^
   193 |
   194 |       if (!groupToUse) {
   195 |         // Skip when no writable group could be determined for the
   current user.
   
=> Your application has errors. Waiting for file change.
```

@StorytellerCZ are you able to run `bash ./packages/test-in-console/run.sh` locally ?

italojs committed:
chore: Remove platform-specific group determination logic from socket file tests.
